mri/reconstructors/self_calibrating.py

```diff
@@ -0,0 +1,180 @@
+        Verbosity level.
+            1 => Print basic debug information
+            5 => Print all initialization information
+            20 => Calculate cost at the end of each iteration.
+                NOTE : This is computationally intensive.
+            30 => Print the debug information of operators if defined by class
+    Note:
+    -----
+    The user is expected to specify the either prox_op or mu to obtain
+    reconstructions, else the above equations lose the regularization terms
+    resulting in inverse transform as solution.
+    The reconstruction in this case proceeds with a warning.
+    """
+
+    def __init__(self, fourier_op, linear_op=None, prox_op=None, mu=0,
+                 gradient_method="synthesis", lips_calc_max_iter=10,
+                 num_check_lips=10, lipschitz_cst=None, kspace_portion=0.1,
+                 smaps_extraction_mode='gridding',
+                 smaps_gridding_method='linear', n_jobs=1, verbose=0):
+        if linear_op is None:
+            linear_op = WaveletN(
+                wavelet_name="sym8",
+                nb_scale=3,
+                dim=len(fourier_op.shape),
+                n_coils=1,
+                verbose=bool(verbose >= 30),
+            )
+        # Ensure that we are in right multichannel config
+        if linear_op.n_coils != 1:
+            raise ValueError("The value of n_coils for linear operation must "
+                             "be 1 for Self-Calibrating reconstruction!")
+        if gradient_method == 'analysis':
+            grad_class = GradSelfCalibrationAnalysis
+        elif gradient_method == 'synthesis':
+            grad_class = GradSelfCalibrationSynthesis
+        super(SelfCalibrationReconstructor, self).__init__(
+            fourier_op=fourier_op,
+            linear_op=linear_op,
+            prox_op=prox_op,
+            mu=mu,
+            gradient_method=gradient_method,
+            grad_class=grad_class,
+            lipschitz_cst=lipschitz_cst,
+            lips_calc_max_iter=lips_calc_max_iter,
+            num_check_lips=num_check_lips,
+            init_gradient_op=False,
+            verbose=verbose,
+        )
+        self.smaps_gridding_method = smaps_gridding_method
+        self.smaps_extraction_mode = smaps_extraction_mode
+        if type(kspace_portion) == float:
+            self.kspace_portion = (kspace_portion,) * \
+                                  self.fourier_op.samples.shape[-1]
+        else:
+            self.kspace_portion = kspace_portion
+        if len(self.kspace_portion) != self.fourier_op.samples.shape[-1]:
+            raise ValueError("The k-space portion size used to estimate the" +
+                             "sensitivity information is not aligned with" +
+                             " the input dimension")
+        self.n_jobs = n_jobs
+
+    def reconstruct(self, kspace_data, optimization_alg='pogm', x_init=None,
+                    num_iterations=100, reinit_grad_op=False, **kwargs):
```

Comment:
Consider adding a `recompute_Smaps` which basically extracts the sensitivity maps from the input data.

Reply:
This is the purpose of reinit_grad_op, please do check the updated codes on this

Reply:
Maybe I missed something, does `reinit_grad_op` recalculate the sensitivity maps or just recompute the Lipschitz constant, if it recalculate the senisitivity maps then the name is misleading if it recomputes the Lipsichitz constant than the name shoudl change to `recompute_lipschitz`

Reply:
It does both. Reinit_grad_op basically re initializes gradient, due to which both Smaps is recomputed and Lipchitz constant is updated.